chainer/dataset/convert.py

```diff
@@ -239,6 +256,10 @@ def __call__(self, batch, device=None, padding=None):
         if device is not self._device:
             raise ValueError('device is different')
 
+        if self.compute_stream is not None:
+            self.event1.synchronize()
```

Comment:
Why do you use synchronize?

Reply:
This synchronization is necessary to ensure that a CPU thread won't break the data of arrays that might be still being used by GPU kernels for previous iteration.

Suppose you are preparing input data for iteration _t_. Considering double-buffering scheme is used here, the array to be used for iteration _t_ was/is also used for iteration _t-2_. If you update/write the array for iteration _t_ while the array still being used by GPU kernels for iteration _t-2_, the results of iteration _t-2_ might get wrong.

How can we avoid this possible data corruption? The safe way is to make sure that all GPU kernels for iteration _t-2_ are finished, and the following codes are for that purpose.

    259        if self.compute_stream is not None:
    260            self.event1.synchronize()
    261            self.event1.record(stream=self.compute_stream)

As you can see, `self.event1.synchronize()` wait for the event that is "recorded" at previous iteration, more specifically, at the beginning of iteration _t-1_, in other word, the end of iteration _t-2_, on "compute stream". With this event synchronization, you can safely update the array.

Reply:
OK, I understood.